llvm/test/CodeGen/X86/insertelement-var-index.ll

```diff
@@ -1019,7 +1019,7 @@ define <2 x i64> @arg_i64_v2i64(<2 x i64> %v, i64 %x, i32 %y) nounwind {
 ; X86AVX2-NEXT:    movl %edx, (%esp,%esi,4)
 ; X86AVX2-NEXT:    vmovaps (%esp), %xmm0
 ; X86AVX2-NEXT:    vmovaps %xmm0, {{[0-9]+}}(%esp)
-; X86AVX2-NEXT:    incl %ecx
+; X86AVX2-NEXT:    orl $1, %ecx
```

Comment:
Why OR is better than INC? I see they have the same TP/ports but OR has one more byte longer and Lat seems longer too in some case?

Reply:
I guess we're missing addlike matching in the inc patterns?

Reply:
CC @goldsteinn who was working on something similar for #83691

Reply:
The freeze might be in the way. I think I saw something odd with the worklist in DAGCombiner for hoisting freeze. I need to recheck.

Reply:
NB: that patch doesn't fix the issue.

Reply:
Sadly my or_is_add patch did not work either. This is the final DAG

```
SelectionDAG has 36 nodes:                                                       
  t0: ch,glue = EntryToken                                                       
                t2: v2i64,ch = CopyFromReg t0, Register:v2i64 %0                 
              t52: ch = store<(store (s128) into %stack.1)> t0, t2, FrameIndex:i32<1>, undef:i32
              t17: i32,ch = load<(load (s32) from %fixed-stack.2)> t0, FrameIndex:i32<-1>, undef:i32
                  t43: i32 = and t42, Constant:i32<3>                            
                t51: i32 = shl t43, Constant:i8<2>                               
              t48: i32 = or disjoint FrameIndex:i32<1>, t51                      
            t46: ch = store<(store (s32))> t52, t17, t48, undef:i32              
          t47: v4i32,ch = load<(load (s128) from %stack.1)> t46, FrameIndex:i32<1>, undef:i32
        t32: ch = store<(store (s128) into %stack.0)> t0, t47, FrameIndex:i32<0>, undef:i32
          t19: i32 = add FrameIndex:i32<-1>, Constant:i32<4>                     
        t20: i32,ch = load<(load (s32) from %fixed-stack.2 + 4)> t0, t19, undef:i32
              t30: i32 = or t42, Constant:i32<1>                                 
            t35: i32 = and t30, Constant:i32<3>                                  
          t55: i32 = shl t35, Constant:i8<2>                                     
        t53: i32 = or disjoint FrameIndex:i32<0>, t55                            
      t38: ch = store<(store (s32))> t32, t20, t53, undef:i32                    
    t56: v2i64,ch = load<(load (s128) from %stack.0)> t38, FrameIndex:i32<0>, undef:i32
  t14: ch,glue = CopyToReg t0, Register:v2i64 $xmm0, t56                         
      t9: i32,ch = load<(load (s32) from %fixed-stack.0)> t0, FrameIndex:i32<-3>, undef:i32
    t24: i32 = shl t9, Constant:i8<1>                                            
  t42: i32 = freeze t24                                                          
  t15: ch = X86ISD::RET_GLUE t14, TargetConstant:i32<0>, Register:v2i64 $xmm0, t14:1
```

t30 is the `or`. It uses the `t42` freeze of `t24`. We should have been able to hoist the freeze above the `t24` since its the only user.

You can see there are 2 uses of t42. Both of those are nodes that a `freeze` was hoisted above. After the first `freeze` hoisted, the `shl` was used by the hoisted `freeze` and one of the other nodes so the `freeze` wasn't the only user of the `shl`. Then we did the second hoist and the second hoisted `freeze` CSEd with the previous hoisted `freeze` and decreased the use count of the `shl` making the `freeze` the only user. We did not revisit the `freeze` after that.